Support for -msvc targets

nymoral opened the pull request:
Support is added in two steps:
* First, this allows passing custom target when creating a config.
* Secondly, this fixes the 'set' command for environment, allowing to overwrite the Rust linker flags, that are gnu specific by default.

nymoral commented on rust_build_utils/rust_utils.py:
```diff
@@ -102,11 +107,11 @@ def clear_env_variables(config):
             "env"
         ].items():
-            if value[1] == "set":
+            if value[1] in "set":
```
Should be reverted

arg0d commented on rust_build_utils/msvc.py:
```diff
@@ -0,0 +1,149 @@
+        "Preview",
+    ]
+    preference_index = (
+        lambda e: edition_preferences.index(e)
```
The lambda is very awkward to read, just defined a local function instead, i.e.
```
def preference_index(e):
    if e in edition_preferences:
        return edition_preferences.index(e)
    else:
        return len(edition_preferences)
```

arg0d commented on rust_build_utils/msvc.py:
```diff
@@ -0,0 +1,149 @@
+        # Because the change happens in a separate process, after it exits the changes made to the env are lost.
+        # We collect the process output (modified environment) and set current environment to those values.
+        # When setting the environment we save the old values so they can be restored after exiting the context.
+        p = subprocess.run(
```
This looks like it should be `subprocess.check_output`.

nymoral replied:
Didn't do this change. The piped output would add noise and it's not needed. The output is parsed after the process exits.

arg0d replied:
Can you try this? `subprocess.check_call(['your_command'], stderr=subprocess.DEVNULL)`

nymoral replied:
https://docs.python.org/3/library/subprocess.html#subprocess.check_call:

> Code needing to capture stdout or stderr should use [run()](https://docs.python.org/3/library/subprocess.html#subprocess.run) instead:

arg0d commented on rust_build_utils/msvc.py:
```diff
@@ -0,0 +1,149 @@
+            check=True,
+            capture_output=True,
+        )
+        for m in re.finditer(r"^([^=]*)=(.*)$", p.stdout.decode("utf-8"), flags=re.M):
```
Why not `^(.*)=(.*)$`?

nymoral replied:
> Why not `^(.*)=(.*)$`?

It would be a greedy math. A line `A=B=C` would produce a variable `A=B` with value `C`, rather than desired `A` with value `B=C`.

arg0d replied:
I see. In that case non greedy `.*?` should work. I'm not against the original regex, but it was hard for me to understand, I had to paste it into regex101.com

nymoral replied:
Hmmm, for me the `[^=]` thing is obvious, while I had to google the `.*?` syntax. Will make the change never the less.

nymoral replied:
Commented what the regex does

arg0d commented on rust_build_utils/rust_utils.py:
```diff
@@ -29,9 +30,13 @@ class CargoConfig:
-        self.rust_target = GLOBAL_CONFIG[self.target_os]["archs"][self.arch][
-            "rust_target"
-        ]
+        if self.rust_target is None or self.rust_target == "":
```
`if not self.rust_target:` is more Pythonic, no?

arg0d commented on rust_build_utils/rust_utils.py:
```diff
@@ -415,8 +434,12 @@ def post_build(project: Project, config: CargoConfig, packages: PackageList) ->
     print("|EXECUTE| {}".format(" ".join(command)))
-    subprocess.check_call(command)
-    print("")
+    p = subprocess.run(command, capture_output=True, text=True)
```
I don't agree with this change.
- This breaks log streaming. If the build takes 5 minutes, you won't see any logs until the build has finished. This also breaks output colouring when used interactively.
- Command output is very useful when debugging CICD. Only printing the output for commands that failed would make it more difficult to understand what is happening under the hood.

nymoral replied:
Yep, was added for debugging and done so incorrectly. Will remove.

jjanowsk commented on rust_build_utils/msvc.py:
```diff
@@ -0,0 +1,149 @@
+    # Sample location of vcvarsall script:
+    # "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvarsall.bat"
+    # We begin by finding microsoft visual studio installation.
+    mvs_path = Path(r"C:\Program Files\Microsoft Visual Studio")
```
The default path in my experience for Visual Studio is "C:\Program Files (x86)\Microsoft Visual Studio". I assume you've tested it on some machine where it was "Program Files" instead, so I guess it's worth checking both options

nymoral replied:
> The default path in my experience for Visual Studio is "C:\Program Files (x86)\Microsoft Visual Studio". I assume you've tested it on some machine where it was "Program Files" instead, so I guess it's worth checking both options

Not for the scripts: https://learn.microsoft.com/en-us/cpp/build/building-on-the-command-line?view=msvc-170#developer_command_file_locations
But to allow full configuration a version of the function with explicit path should probably be provided as well

jjanowsk replied:
It's most likely a bug in their documentation :) Few lines below that they have this:
```
Create your own command prompt shortcut
Open the Properties dialog for a developer command prompt shortcut to see the command target used. For example, the target for the x64 Native Tools Command Prompt for VS 2019 shortcut is something similar to:

%comspec% /k "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat"
```
But  for example on GitHub runners it's "Program Files" so in general those two should be checked

nymoral replied:
Yes I will add a check for both. I guess this is very much version dependent, because 2022 version is installed in `Program Files` on our runners :shrug:

jjanowsk commented on rust_build_utils/msvc.py:
```diff
@@ -0,0 +1,149 @@
+
+
+@contextmanager
+def activate_msvc(
```
In general I'm not sure this should be a reponsibility of a build script to prepare an environment like this. It is machine dependent and I also feel like one should be able to pick version they want to use instead of relying on some arbitrary ordering. But I see a value which is just an ease of use for someone who is not used to working with msvc. Still I'm not sure if this is enough of an argument for me

nymoral replied:
> In general I'm not sure this should be a reponsibility of a build script to prepare an environment like this. It is machine dependent and I also feel like one should be able to pick version they want to use instead of relying on some arbitrary ordering. But I see a value which is just an ease of use for someone who is not used to working with msvc. Still I'm not sure if this is enough of an argument for me

Ordering can be changed based on `*_preference` parameters. As far as responsibilities go I'm very conflicted myself. Ideally any user of build utils would activate the msvc environment before calling the build functions. But the environment is required only for specific Rust targets, meaning that the API of build scripts would read "you need to call `activate_msvc` before calling `cargo_build` if `config.target` is "-msvc" - it would introduce a required function call for some parameters, meaning those are not really independent parameters. Secondly, our build scripts do make a lot of assumptions about where they will run - firstly that Rust is installed and added to the path, secondly about the version of Rust used (and NDK version in the case of android scripts). Their purpose is not be a general purpose scripts (for that the raw tools are the only complete fit), but rather abstract away *our* infrastructure, which does include specific machines.

arg0d replied:
I think this script for detecting and sourcing VS vars is great. Its always a pain on Windows to find the correct path to these `vars` scripts. I guess it would be great if it was possible to somehow override this path on the runner instance by adding some env, in case the runner has multiple VS installations and the default ordering in the script is not sufficient. But I think this is an edge case, and picking the latest installed version is a great default.

jjanowsk replied:
> Ordering can be changed based on *_preference parameters
It can be changed but it's not exposed in public API of rust_utils. So if I just run `cargo_build` from my project, I don't have a way to influence this choice.

> API of build scripts would read "you need to call activate_msvc before calling cargo_build
I don't think this is an API thing. I do not propose to ask users to call activate_msvc  themselves in the script. I just say that maybe it should not be a part of this script at all. Like you said in all other instances we assume environment is correctly configured. e.g. rust is always in a path, if building for android android ndk is in the path. Why should we treat msvc differently instead of assuming that cl.exe, link.exe etc are in the path before someone called the script. I would simply check if they are and if not I would inform the user that if they want to use msvc toochain they need to have msvc tools in path.
Maybe it could be also done this way: if the tools are not set then use activate_msvc. If the user have msvc tools in the path then skip it. This way activate_msvc can be simplified. No need to allow user to influence it's choices. For 99% of them the default is ok, and the rest can just set the environment themselves.

jjanowsk replied:
> that doesn't stop you from using msvc module individually

I think in this form it does stop me, because even if I set the env myself, when I run `cargo_build` with `-msvc` target, the `activate_msvc` function will be called anyway, so if there is some issue with this function I don't have an easy way to skip this step. That's why I would propose to add some check that if env is set already do not execute this function, or I would remove this step entirely (That's actually my opinion. That the only thing this PR should do is to pass target argument to the cargo invocation)

> the whole if target.ends_with("-msvc"): activate_msvc() will go away

I think I don't get this part. What is stopping you currently from removing this part? What would the API that would allow this?

arg0d replied:
I just realized that automatically activating the environment is especially nice for local use. Building libtelio on Windows machine fails without explicitly sourcing or using the special VS pre-sourced terminal.

jjanowsk replied:
Yeah, it's convenient but it might be annoying if it fails. But now there is `is_msvc_active()` function which allows users to initialize msvc toolchain on their own before hand and the script will not fail, so I'm OK with it now.

jjanowsk commented:
Please add building with msvc target to the CI

jjanowsk commented on rust_build_utils/rust_utils.py:
```diff
@@ -129,7 +134,7 @@ def config_local_env_vars(config, local_config):
                 GLOBAL_CONFIG[config.target_os]["env"] = {env: tuple}
                 return
-            if env in GLOBAL_CONFIG[config.target_os]["env"]:
+            if env in GLOBAL_CONFIG[config.target_os]["env"] and tuple[1] != "set":
```
maybe `tuple[1] != "append"` would be more self explanatory

jjanowsk commented on rust_build_utils/rust_utils.py:
```diff
@@ -129,7 +134,7 @@ def config_local_env_vars(config, local_config):
             if not "env" in GLOBAL_CONFIG[config.target_os]:
                 GLOBAL_CONFIG[config.target_os]["env"] = {env: tuple}
                 return
```
also this is not related, but I suspect this return is a bug. Would you be willing to write UTs for this functions since you're doing fixes here?

nymoral commented:
> Shouldn't we add the target to `rust_utils_config.py`?
> 
> ```
>              "x86_64": {
>                  "rust_target": "x86_64-pc-windows-gnu",
>              },
> +            "aarch64": {
> +                "rust_target": "aarch64-pc-windows-msvc",
> +            },
> ```

Added. And changed  the x86_64 one to -msvc as well

arg0d commented on rust_build_utils/msvc.py:
```diff
@@ -39,60 +42,74 @@ def activate_msvc(
+                mvs_paths
             )
         )
+    print(mvs_versions)
```
leftover debug print?

arg0d commented on rust_build_utils/msvc.py:
```diff
@@ -39,60 +42,74 @@ def activate_msvc(
-            for v in mvs_path.iterdir()
-            if v.is_dir()
-            and (version_preference is None or v.name == version_preference)
+            for v in itertools.chain.from_iterable(p.iterdir() for p in mvs_paths)
```
This is one gnarly way to write a double nested loop :D At this point I'm having trouble to comprehend whats happening here.

arg0d replied:
Maybe the checking part could be separated into a local function? But this way the ifs are inverted as guard clauses, so it makes it hard to read as well :/
```
def check_version(v):
  if not v.is_dir():
    return None
 ..
```

arg0d commented on rust_build_utils/rust_utils_config.py:
```diff
@@ -66,11 +66,11 @@
         "archs": {
             "x86_64": {
-                "rust_target": "x86_64-pc-windows-gnu",
+                "rust_target": "x86_64-pc-windows-msvc",
```
This is a breaking change for anyone cross compiling on Linux. At the moment I see that Libdrop is cross compiled from Linux. This is not a blocker, but should be mentioned in CHANGELOG.md, and major version should be bumped.

nymoral replied:
There's no changelog file here, and we agreed with tech leads to do this breaking change. ARM64 support will require MSVC toolchain and a windows runner anyways

arg0d replied:
So create `CHANGELOG.md` and mention that starting with version `X.X.X` cross compiling Windows is not supported anymore and compiling for Windows has to be done on Windows host.

arg0d reviewed (changes requested):
IMO the double nested loop should be made easier to read, and CHANGELOG.md should be updated to mention the breaking change for Windows `rust_target`.

jjanowsk reviewed (commented):
Looks good more or less. IMO activate_msvc could just activate the latest version (random edition). It should not matter. If someone needs a specific one they can activate it beforehand instead of playing with version_preference and edition_preference. In that case activate_msvc could be much simplified. Just my extra 2 cents.

jjanowsk commented on rust_build_utils/rust_utils_config.py:
```diff
@@ -66,11 +66,11 @@
             },
-        },
-        "env": {
-            "RUSTFLAGS": ([" -C link-arg=-s "], "set"),
```
This should also be mentioned in a changelog since some libs can rely on this default being here

nymoral replied:
This is incompatible with msvc compiler

jjanowsk commented:
and please, add CI job to test that both msvc and gnu toolchains work for Windows

LukasPukenis commented:
Just wondering, what's the stage of this PR? I tried it and it seems to work for Libdrop just by pointing to this pull request and running on windows runner
